<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 4</b> 
    
Great, Vivian! You've done a great job and now your project has been accepted.
    
Thank you for your work and I wish you success in the following projects!

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 3</b> 
    
Vivian, hello!
    
Please **delete all duplicates in the dataset**, a more detailed comment on how and where to do this is marked in red and v 3.0

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 2</b> 
    
You did a good job on the comments, but I still have one more comment, it is marked v 2.0 and red. Please fix it and I will accept your project)

<div style="border:solid blue 2px; padding: 20px">
  
**Hello Vivian**

I am happy to review your project today!

You will find my comments in coloured cells marked as 'Reviewer's comment'. The cell colour will vary based on the contents - I am explaining it further below. 

**Note:** Please do not remove or change my comments - they will help me in my future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

You are also very welcome to leave your comments / describe the corrections you've done / ask me questions, marking them with a different colour. You can use the example below: 

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>

## Basic Python - Project <a id='intro'></a>

<div style="border:solid green 2px; padding: 20px">
    
<div class="alert alert-success">
<b>Review summary</b> 
    
Vivian, thanks for submitting the project. You've done a very good job and I enjoyed reviewing it.

    
There are only a few critical comments that need to be corrected. You will find them in the red-colored cells in relevant sections. If you have any questions please write them when you return your project. 
    
I'll be looking forward to getting your updated notebook.

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 
    
Before sending a new version, please click `Kernel - > Restart & Run all` to make sure that your code is working

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Title and introduction are essential parts of the project. Make sure you do not forget to include it in your further projects. 
    
It is optimal if introduction consists of:
    
- brief description of the situation;
- goal of the project;
- description of the data we are going to use.
</div>


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [1]:
# importing pandas
import pandas as pd 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> Needed library has been added </div>

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [2]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> The correct path to the file is specified: the slash at the beginning of the path is very important, as it indicates that you need to search for the file in the root folder. </div>

Print the first 10 table rows:

In [3]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [4]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great - you've used a comprehensive set of methods to have a first look at the data.
    

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Please note that it is highly recommended to add a conclusion / summary after each section and describe briefly your observations and / or major outcomes of the analysis.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [5]:
# the list of column names in the df table
list(df.columns)

['   name',
 'Character',
 'r0le',
 'TITLE',
 '  Type',
 'release Year',
 'genres',
 'imdb sc0re',
 'imdb v0tes']

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [6]:
# renaming columns
df = df.rename(columns={
    '   name': 'name',
    'Character': 'character',
    'r0le': 'role',
    'TITLE': 'title',
    '  Type': 'type',
    'release Year': 'release_year',
    'imdb sc0re': 'imdb_score',
    'imdb v0tes': 'imdb_votes'
})
    

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
This is a good way to rename the columns.

Check the result. Print the names of the columns once more:

In [7]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imdb_score', 'imdb_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [8]:
# calculating missing values
df.isna().sum()

name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imdb_score      4609
imdb_votes      4726
dtype: int64

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
The isna() method is selected to find the missing values, it's great!

We identified missing values in several columns. While the missing value in `'title'` isn't critical, missing values in `'imdb_score'` and `'imdb_votes'` affect around 6% of the data, which could impact our analysis. To ensure data integrity, we'll drop all rows with missing values.

In [9]:
# dropping rows where columns with scores, and votes have missing values
df = df.dropna(subset=['imdb_score', 'imdb_votes'])

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 
    
Please note that we have to remove the missing values in the columns, but not the entire columns)

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Perfect!

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [10]:
# counting missing values
df.isna().sum()

name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imdb_score      0
imdb_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [11]:
# counting duplicate rows
duplicates = df.duplicated()
duplicates.sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [12]:
# Produce table with duplicates (with original rows included) and review last 5 rows

duplicated_rows = df[df.duplicated()]
duplicated_rows.tail()

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imdb_score  imdb_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [13]:
# removing duplicate rows
df.drop_duplicates()

name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85573       A??da Morales                  Maritza     ACTOR      Lokillo   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes  
0          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
1          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
2          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
3          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
4          MOVIE          1976  ['drama', 'crime']         8.2    808582.0  
...          ...           ...                 ...         ...         ...  
85573  the movie          2021          ['comedy']         3.8        68.0  
85574  the movie          2021          ['comedy']         3.8        68.0  
85575  the movie          2021          ['comedy']         3.8        68.0  
85576  the movie          2021          ['comedy']         3.8        68.0  
85578  the movie          2021          ['comedy']         3.8        68.0  

[73859 rows x 9 columns]

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 
    
Not quite right. We just needed to drop all the duplicates in the `df` dataset itself, like:
    
    df = df.drop_duplicates()

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 2</b> 
    
Unfortunately, the duplicates are still not deleted, please use the code above to delete duplicates in the dataset

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 3</b> 
    
Vivian, please pay attention to the code cell below, now all the 6994 remain in the dataset, but we should **remove all of them**. Please us this code to drop all the duplicates:
    
    df = df.drop_duplicates()

Check for duplicate rows once more to make sure you have removed all of them:

In [14]:
# checking for duplicates
df = df.drop_duplicates()

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [15]:
# viewing unique type names
df['type'].sort_values().unique()

array(['MOVIE', 'SHOW', 'movies', 'shows', 'the movie', 'tv', 'tv series',
       'tv show', 'tv shows'], dtype=object)

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

Please note, that according to the technical task it was asked:
    
* For the **sorted** column, call the method that will return all unique column values
    
So, we should use sorting here

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Great!

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

Please pay attention, that according to the technical task it was asked:
    
* declare the **function** replace_wrong_show() with two parameters
    
So, we should create a **function** here

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Very well!

In [16]:
# function for replacing implicit duplicates
wrong_show_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'SHOW'

def replace_wrong_show(df, wrong_shows_list, correct_show):
    for wrong_show in wrong_show_list:
        df.loc[df['type'] == wrong_show, 'type'] = correct_show 
    return df
    
movies_and_shows = replace_wrong_show(df, wrong_show_list, correct_show)

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [17]:
# removing implicit duplicates
print(movies_and_shows)

                     name                character      role        title  \
0          Robert De Niro            Travis Bickle     ACTOR  Taxi Driver   
1            Jodie Foster            Iris Steensma     ACTOR  Taxi Driver   
2           Albert Brooks                      Tom     ACTOR  Taxi Driver   
3           Harvey Keitel  Matthew 'Sport' Higgins     ACTOR  Taxi Driver   
4         Cybill Shepherd                    Betsy     ACTOR  Taxi Driver   
...                   ...                      ...       ...          ...   
85573       A??da Morales                  Maritza     ACTOR      Lokillo   
85574    Adelaida Buscato               Mar??a Paz     ACTOR      Lokillo   
85575  Luz Stella Luengas             Karen Bayona     ACTOR      Lokillo   
85576        In??s Prieto                    Fanny     ACTOR      Lokillo   
85578      Julian Gaviria                  unknown  DIRECTOR      Lokillo   

            type  release_year              genres  imdb_score  imdb_votes 

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [18]:
# viewing unique genre names
df['type'].sort_values().unique()

array(['MOVIE', 'SHOW', 'movies', 'the movie'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 
    
Unfortunately, the code below cannot be executed and it crashes with an error. Please check its correctness

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 

Now the code works as it should!

In [19]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)
# give the slice of dataframe new name

filtered = df[(df['release_year'] >= 1999) & (df['type'] == 'SHOW')]
filtered

name                character      role      title  \
1664           Jeff Probst           Himself - Host     ACTOR   Survivor   
2076         Mayumi Tanaka  Monkey D. Luffy (voice)     ACTOR  One Piece   
2077          Kazuya Nakai     Roronoa Zoro (voice)     ACTOR  One Piece   
2078         Akemi Okamura             Nami (voice)     ACTOR  One Piece   
2079      Kappei Yamaguchi            Usopp (voice)     ACTOR  One Piece   
...                    ...                      ...       ...        ...   
85433     Maneerat Kam-Uan                       Ae     ACTOR  Let's Eat   
85434    Rudklao Amratisha                  unknown     ACTOR  Let's Eat   
85435      Jaturong Mokjok                  unknown     ACTOR  Let's Eat   
85436     Pisamai Wilaisak                  unknown     ACTOR  Let's Eat   
85437  Sarawut Wichiensarn                  unknown  DIRECTOR  Let's Eat   

       type  release_year                                             genres  \
1664   SHOW          2000                                        ['reality']   
2076   SHOW          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2077   SHOW          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2078   SHOW          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
2079   SHOW          1999  ['animation', 'action', 'comedy', 'drama', 'fa...   
...     ...           ...                                                ...   
85433  SHOW          2021                                ['drama', 'comedy']   
85434  SHOW          2021                                ['drama', 'comedy']   
85435  SHOW          2021                                ['drama', 'comedy']   
85436  SHOW          2021                                ['drama', 'comedy']   
85437  SHOW          2021                                ['drama', 'comedy']   

       imdb_score  imdb_votes  
1664          7.4     24687.0  
2076          8.8    117129.0  
2077          8.8    117129.0  
2078          8.8    117129.0  
2079          8.8    117129.0  
...           ...         ...  
85433         8.2         5.0  
85434         8.2         5.0  
85435         8.2         5.0  
85436         8.2         5.0  
85437         8.2         5.0  

[13430 rows x 9 columns]

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [20]:
# rounding column with scores
filtered['imdb_score'] = filtered['imdb_score'].round()

#checking the outcome with tail()
filtered.tail()

/tmp/ipykernel_51/2790141854.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered['imdb_score'] = filtered['imdb_score'].round()


name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imdb_score  imdb_votes  
85433  ['drama', 'comedy']         8.0         5.0  
85434  ['drama', 'comedy']         8.0         5.0  
85435  ['drama', 'comedy']         8.0         5.0  
85436  ['drama', 'comedy']         8.0         5.0  
85437  ['drama', 'comedy']         8.0         5.0

It is now time to identify outliers based on the number of votes.

In [21]:
# Use groupby() for scores and count all unique values in each group, print the result
scores = filtered.groupby('imdb_score').size()
scores

imdb_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
dtype: int64

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
The dataframe was filtered and grouped flawlessly, but the result will change after removing the duplicates

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [22]:
# filter dataframe using two conditions (scores to be in the range 4-9)
scores = filtered[(filtered['imdb_score'] >= 4) & (filtered['imdb_score'] <= 9)]

# group scores and corresponding average number of votes, reset index and print the result
average_votes = scores.groupby('imdb_score').mean().reset_index()
average_votes

imdb_score  release_year     imdb_votes
0         4.0   2018.072222    5277.583333
1         5.0   2018.881757    3143.942568
2         6.0   2018.627506    3481.717322
3         7.0   2018.061411    8727.068211
4         8.0   2017.120611   30299.460967
5         9.0   2015.287253  126904.109515

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [23]:
# round column with averages
average_votes = average_votes.round(0)

# rename columns
average_votes.rename(columns={'imdb_votes': 'average_votes'}, inplace=True)

# print dataframe in descending order
average_votes = average_votes.sort_values(by='average_votes', ascending=False)
average_votes

imdb_score  release_year  average_votes
5         9.0        2015.0       126904.0
4         8.0        2017.0        30299.0
3         7.0        2018.0         8727.0
0         4.0        2018.0         5278.0
2         6.0        2019.0         3482.0
1         5.0        2019.0         3144.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Great! Also correct rounding and grouping in this section

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.